### PR TITLE
[BEAM-2899] Add a Bare-bones ReferenceRunner Job Service

### PR DESCRIPTION
**Pull request by tgroh**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
This will eventually accept Job API calls and execute a Pipeline using
the ReferenceRunner backend.

This change exists primarily to create the appropriate module and POM.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @lukecwik

**Comment by tgroh** [comment]
The jenkins failure appears unrelated, but persistent

java.lang.NoClassDefFoundError: Could not initialize class org.xerial.snappy.Snappy
	at org.xerial.snappy.SnappyOutputStream.<init>(SnappyOutputStream.java:97)
	at org.xerial.snappy.SnappyOutputStream.<init>(SnappyOutputStream.java:89)
	at org.xerial.snappy.SnappyOutputStream.<init>(SnappyOutputStream.java:79)
	at org.apache.beam.sdk.util.SerializableUtils.serializeToByteArray(SerializableUtils.java:48)

I don't believe  this change could cause the examples to fail  (and in fact it did not, as it is later in the reactor)

**Comment by tgroh** [comment]
retest this please

**Comment by tgroh** [comment]
And then a Javadoc failure. Jenkins has been very temperamental today

File: runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobServer.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+package org.apache.beam.runners.reference.job;
+
+import io.grpc.Server;
+import io.grpc.ServerBuilder;
+import java.io.IOException;
+
+/** A program that runs a {@link ReferenceRunnerJobService}. */
+public class ReferenceRunnerJobServer {
+  public static void main(String[] args) throws IOException, InterruptedException {
+    if (args.length != 1) {
```
Comment (lukecwik) [inline runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobServer.java]: Consider using args4j, they support shiny annotations or any other args library.

File: runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobService.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.beam.sdk.common.runner.v1.JobApi.GetJobStateResponse;
+import org.apache.beam.sdk.common.runner.v1.JobApi.PrepareJobResponse;
+import org.apache.beam.sdk.common.runner.v1.JobApi.RunJobRequest;
+import org.apache.beam.sdk.common.runner.v1.JobServiceGrpc.JobServiceImplBase;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link JobServiceImplBase Job Service} which implements methods by executing with the
+ * ReferenceRunner. The ReferenceRunner uses containers over the Fn API to execute a Pipeline on
```
Comment (lukecwik) [inline runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobService.java]: The ReferenceRunner uses the portability framework to execute a Pipeline on a single machine.

File: runners/reference/pom.xml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  </parent>
+
+  <artifactId>beam-runners-reference</artifactId>
+
+  <name>Apache Beam :: Runners :: Reference</name>
+
+  <packaging>jar</packaging>
+
+  <build>
+    <resources>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: You have no resources, no point of including this statement.

File: runners/reference/pom.xml [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    </dependency>
+
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+    </dependency>
+
+    <!-- build dependencies -->
+    <dependency>
+      <groupId>com.google.auto.value</groupId>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: add it when we need it.

File: runners/reference/pom.xml [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      <artifactId>auto-value</artifactId>
+      <scope>provided</scope>
+    </dependency>
+
+    <!-- test dependencies -->
+    <dependency>
+      <groupId>org.hamcrest</groupId>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: drop

File: runners/reference/pom.xml [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+      <artifactId>hamcrest-all</artifactId>
+      <scope>provided</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>provided</scope>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: scope -> test

File: runners/reference/pom.xml [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-jdk14</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-all</artifactId>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: drop

File: runners/reference/pom.xml [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>beam-sdks-java-core</artifactId>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: drop

File: runners/reference/pom.xml [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>beam-runners-core-java</artifactId>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: drop

File: runners/reference/pom.xml [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
+
+    <!-- required for XMLSourceTest -->
+    <dependency>
+      <groupId>org.codehaus.woodstox</groupId>
+      <artifactId>stax2-api</artifactId>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: drop

File: runners/reference/pom.xml [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+      <version>${stax2.version}</version>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.codehaus.woodstox</groupId>
+      <artifactId>woodstox-core-asl</artifactId>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: drop

File: runners/reference/pom.xml [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+          <groupId>javax.xml.stream</groupId>
+          <artifactId>stax-api</artifactId>
+        </exclusion>
+      </exclusions>
+      <scope>test</scope>
+    </dependency>
+
+    <!-- needed for eclipse-jdt generated core as the test-jar references classes from this -->
+    <dependency>
+        <groupId>com.google.cloud.dataflow</groupId>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: drop

File: runners/reference/pom.xml [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        <artifactId>google-cloud-dataflow-java-proto-library-all</artifactId>
+        <scope>test</scope>
+    </dependency>
+
+    <!-- transitive test dependencies from beam-sdk-java-core -->
+    <dependency>
+      <groupId>com.fasterxml.jackson.dataformat</groupId>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: drop

File: runners/reference/pom.xml [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+      <artifactId>jackson-dataformat-yaml</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>beam-sdks-common-fn-api</artifactId>
```
Comment (lukecwik) [inline runners/reference/pom.xml]: drop

File: runners/reference/pom.xml [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+  <parent>
+    <groupId>org.apache.beam</groupId>
+    <artifactId>beam-runners-parent</artifactId>
+    <version>2.2.0-SNAPSHOT</version>
+    <relativePath>../pom.xml</relativePath>
+  </parent>
+
+  <artifactId>beam-runners-reference</artifactId>
+
```
Comment (lukecwik) [inline runners/reference/pom.xml]: remove new line

File: runners/reference/pom.xml [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+  <name>Apache Beam :: Runners :: Reference</name>
+
```
Comment (lukecwik) [inline runners/reference/pom.xml]: Add description

File: runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobServer.java [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+
+package org.apache.beam.runners.reference.job;
+
+import io.grpc.Server;
+import io.grpc.ServerBuilder;
+import java.io.IOException;
+
+/** A program that runs a {@link ReferenceRunnerJobService}. */
+public class ReferenceRunnerJobServer {
+  public static void main(String[] args) throws IOException, InterruptedException {
```
Comment (lukecwik) [inline runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobServer.java]: Separate parsing the args from launching the service so you can call a method with the parameters already. Will help with future testing.

Comment (tgroh) [reply runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobServer.java]: Done

File: runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobServer.java (hunk 1 above)
Comment (tgroh) [reply runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobServer.java]: Done.

File: runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobService.java (hunk 2 above)
Comment (tgroh) [reply runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobService.java]: Done.

File: runners/reference/pom.xml (hunk 15 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 16 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 3 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 4 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 5 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 6 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 7 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 8 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 9 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 11 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 13 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 12 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/pom.xml (hunk 14 above)
Comment (tgroh) [reply runners/reference/pom.xml]: Done.

File: runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobServer.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+import org.kohsuke.args4j.CmdLineParser;
+import org.kohsuke.args4j.Option;
+
+/** A program that runs a {@link ReferenceRunnerJobService}. */
+public class ReferenceRunnerJobServer {
+  public static void main(String[] args) throws IOException, InterruptedException {
+    ServerConfiguration configuration = new ServerConfiguration();
+    try {
+      new CmdLineParser(configuration).parseArgument(args);
+    } catch (CmdLineException e) {
```
Comment (lukecwik) [inline runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobServer.java]: Follow this for printing out usage/help:
https://github.com/kohsuke/args4j/blob/master/args4j/examples/SampleMain.java#L70

Comment (tgroh) [reply runners/reference/src/main/java/org/apache/beam/runners/reference/job/ReferenceRunnerJobServer.java]: Done.

**Comment by tgroh** [comment]
Run Precommit Pipeline Java only

**Comment by tgroh** [comment]
Run Precommit Pipeline java

**Comment by lukecwik** [comment]
retest this please

**Comment by lukecwik** [comment]
LGTM, merge once green

**Comment by tgroh** [comment]
retest this please

**Pull request closed without merging**